sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/cpu/CpuMonitor.java

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.cpu;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.ref.WeakReference;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Monitors history of CPU consumption. This is a singleton class and can support multiple cosmos clients.
+ *
+ * is used for tracking multiple cosmos clients registers to this CPU monitor.
+ * in the absence of a listener the CpuMonitor will shutdown.
+ */
+public class CpuMonitor {
+    private final static int DEFAULT_REFRESH_INTERVAL_IN_SECONDS = 10;
+    private final static int HISTORY_LENGTH = 6;
```

Comment:
I know these are the same values we use in .Net - IMO shorter intervals would have been good. Wondering whether we should start with 1 second when introducing this in Java?

Reply:
CPU reading is expensive. so I would rather we don't do it every second. that's very aggressive.

Reply:
From my tests (on .Net) it wouldn't have been a performance problem - especially because in Java we do it for every single TransportException (or requesting request diagnostics) right now. The problem I have with 10 seconds is that timeouts for customers using very aggressive request timeouts are hard to pinpoint to temporarily high CPU usage. 10 seconds is decent - and for .Net covers the vast majority of cases. So not blocking on this. I just think especially when we talk about collecting the metrics for the telemetry workstream Naveen is working on taking CU samples once per second is more useful. Reporting could be constrained to the top 10 values within the last x samples (to keep coverage over the 60 seconds interval in the CPU history text we emit today. But anyway - for this change I would understand if you go with the save assumption of using the same values as in .Net - just don't think that will be sufficient long term.

Reply:
the current behaviour on master: In Java we don't do for every single TransportException by default. CPU capture will happen only if the end user explicitly invokes toString on the diagnostics of the request.

I will lower it to 5 seconds. the middle ground between 1 and 10 seconds :-)

Reply:
lowered to 5 seconds. thanks for the conversation.